nodestream/pipeline/pipeline.py

```diff
@@ -1,280 +1,195 @@
-                self.errors.append(return_state)
-
-        # Raise the error and log it.
-        if self.errors:
-            raise PipelineException(self.errors)
-
-    async def run(
-        self, progress_reporter: Optional[PipelineProgressReporter] = None
-    ) -> AsyncGenerator[Any, Any]:
-        self.logger.info("Starting Pipeline")
-        tasks = self.build_steps_into_tasks(progress_reporter)
-        return_states = await asyncio.gather(*tasks, return_exceptions=True)
-        self.propogate_errors_from_return_states(return_states)
-        self.logger.info("Pipeline Completed")
 
     def get_child_expanders(self) -> Iterable[ExpandsSchema]:
         return (s for s in self.steps if isinstance(s, ExpandsSchema))
+
+    async def run(self, reporter: PipelineProgressReporter):
+        """Run the pipeline.
+
+        This method is used to run the pipeline. It will start, stop, and run
+        the steps in the pipeline in sequence. It will pass records between the
+        steps in the pipeline using channels. The pipeline will report on the
+        progress of the pipeline using the `PipelineProgressReporter`. The
+        pipeline will run in an asynchronous context. The pipeline will block
+        until all steps in the pipeline are finished.
+
+        This method does not return anything. If an error occurs during the
+        processing of the pipeline, it will be reported using the
+        `PipelineProgressReporter`.
+
+        Args:
+            channel_size: The size of the channels used to pass records between
+                steps in the pipeline.
+            reporter: The `PipelineProgressReporter` used to report on the
+                progress of the pipeline.
+        """
+        # Create the input and output channels for the pipeline. The input
+        # channel is used to pass records from the previous step to the current
+        # step. The output channel is used to pass records from the current
+        # step to the next step. The channels are used to pass records between
+        # the steps in the pipeline. The channels have a fixed size to control
+        # the flow of records between the steps.
+        executors: List[StepExecutor] = []
+        current_input, current_output = channel(self.step_outbox_size)
+        pipeline_output = PielineOutput(current_input, reporter)
+
+        # Create the executors for the steps in the pipeline. The executors
+        # will be used to run the steps concurrently. The steps are created in
+        # reverse order so that the output of each step is connected to the
+        # input of the next step.
+        for index, step in reversed(list(enumerate(self.steps))):
+            context = StepContext(index, reporter)
+            current_input, next_output = channel(self.step_outbox_size)
+            exec = StepExecutor(step, current_input, current_output, context)
+            current_output = next_output
+            executors.append(exec)
```

Comment:
Just a thought about this one, I'm assuming this implies that the context for each step will be initialized in reverse order (Extractor 3, Transformer 2, Ingester 1, Writer 0). It kind of seems odd in terms of identifying the ordering of events when looking at the context logs. I believe this piece can be unreversed with the pipeline output being initialized at the end. Also I think this is a good oppertunity for the StepContext to obtain metadata about the step name from the class info here.

Reply:
Yeah thats miss on my part. `Extractor` from your example should be `0`. I'll math that.